### PR TITLE
Fix compilation on LLVM 5.0

### PR DESCRIPTION
**Pull request by yuyichao**
Ref https://github.com/llvm-mirror/llvm/commit/ee809acd59b2d357c37ac1276aabb6620f94cb7b
Ref https://github.com/llvm-mirror/llvm/commit/966d9ebd7f825b6272aa691dcef6baada1129c5c
Ref https://github.com/llvm-mirror/llvm/commit/a81793582b3c47869680d354a97d59c55779c349
Ref https://github.com/llvm-mirror/llvm/commit/72786af0ac80b7d63f296d5cc8045e56535461ce
Ref https://github.com/llvm-mirror/llvm/commit/67191c1ced59d7790d8423d3a06b986baf0b9256
Ref https://github.com/llvm-mirror/llvm/commit/08ef6db99166c2f37a05c252152f352cbe2f481b

Tested on LLVM SVN so there could be 6.0 changes so I'll need to check that.
Will also need https://reviews.llvm.org/D34335

Replaces https://github.com/JuliaLang/julia/pull/22971

### CODE REVIEW

**Comment by yuyichao** [comment]
More failures when compiled with LLVM svn
```
Invalid derived pointer in jlcall
          %267 = call cc37 %jl_value_t addrspace(10)* @jl_f_svec(%jl_value_t addrspace(10)* addrspacecast (%jl_value_t* null to %jl_value_t addrspace(10)*), %jl_value_t addrspace(10)** null, i32 0), !dbg !15659
Invalid derived pointer in jlcall
          %267 = call cc37 %jl_value_t addrspace(10)* @jl_f_svec(%jl_value_t addrspace(10)* addrspacecast (%jl_value_t* null to %jl_value_t addrspace(10)*), %jl_value_t addrspace(10)** null, i32 0), !dbg !15659
```
This particular one looks harmless enough?

**Comment by Keno** [comment]
No, that call is wrong. `cc37` is pre jlcall ABI lowering, so that call should be:
```
call cc37 %jl_value_t addrspace(10)* @jl_f_svec(%jl_value_t addrspace(10)* addrspacecast (%jl_value_t* null to %jl_value_t addrspace(10)*)),
```

**Comment by yuyichao** [comment]
Ah, yeah. Somehow the jlcall lowering didn't change the CC???? That's strange..

**Comment by yuyichao** [comment]
Checked that LLVM 5.0 C++ code compilation also passed. There's one change that's only in SVN and a few other version check typos.....

**Comment by yuyichao** [comment]
Seems that instcombine transformed
```llvm
call cc37 %jl_value_t addrspace(10)* bitcast (%jl_value_t addrspace(10)* (%jl_value_t addrspace(10)*, %jl_value_t addrspace(10)**, i32)* @jl_f_svec to %jl_value_t addrspace(10)* (%jl_value_t addrspace(10)*)*)(%jl_value_t addrspace(10)* addrspacecast (%jl_value_t* null to %jl_value_t addrspace(10)*))
```

to

```llvm
call %jl_value_t addrspace(10)* @jl_f_svec(%jl_value_t addrspace(10)* addrspacecast (%jl_value_t* null to %jl_value_t addrspace(10)*), %jl_value_t addrspace(10)** null, i32 0)
```

in `llvm::InstCombiner::transformConstExprCastCall`

File: src/codegen.cpp [hunk 1]
```diff
@@ -1288,8 +1289,15 @@ uint64_t jl_get_llvm_fptr(void *function)
 {
     Function *F = (Function*)function;
     uint64_t addr = getAddressForFunction(F->getName());
-    if (!addr)
+    if (!addr) {
+#if JL_LLVM_VERSION >= 50000
+        if (auto exp_addr = jl_ExecutionEngine->findUnmangledSymbol(F->getName()).getAddress()) {
+            addr = exp_addr.get();
+        }
```
Comment (singam-sanjay) [inline src/codegen.cpp]: What happens when exp_addr is NULL ?

File: src/codegen.cpp [hunk 2]
```diff
@@ -1288,8 +1289,15 @@ uint64_t jl_get_llvm_fptr(void *function)
 {
     Function *F = (Function*)function;
     uint64_t addr = getAddressForFunction(F->getName());
-    if (!addr)
+    if (!addr) {
+#if JL_LLVM_VERSION >= 50000
+        if (auto exp_addr = jl_ExecutionEngine->findUnmangledSymbol(F->getName()).getAddress()) {
+            addr = exp_addr.get();
```
Comment (singam-sanjay) [inline src/codegen.cpp]: Will you be handling the errors as well ?

**Comment by yuyichao** [comment]
Worked around the instcombine issue by marking the function as `thunk`. It's a workaround though it's also not too wrong in semantics. I'll let @Keno decide if this is an llvm bug or not.

The only failing test is due to https://github.com/JuliaLang/julia/blob/1ce3c8e60e82fffccfbd65d5d8a58c9314571915/src/llvm-alloc-opt.cpp#L391 Will probably take care of it in a different PR.....

**Comment by singam-sanjay** [comment]
@yuyichao are "broken tests" in Julia similar to "unsupported tests" in LLVM ? context : https://travis-ci.org/JuliaLang/julia/jobs/260298543

**Comment by tkelman** [comment]
"broken test" means should pass but doesn't, so if it starts passing, that means a bug was fixed

**Comment by singam-sanjay** [comment]
@tkelman thanks for letting me know !

@yuyichao https://github.com/llvm-mirror/llvm/commit/9aafb854cc7cb8df8338c50cb411a54ce1e09796 is causing the build to fail at src/disasm.cpp

```
julia_src/src/disasm.cpp: In function ‘void jl_dump_asm_internal(uintptr_t, size_t, int64_t, const llvm::object::ObjectFile*, llvm::DIContext*, llvm::raw_ostream&, const char*)’:
julia_src/src/disasm.cpp:665:32: error: ‘Default’ is not a member of ‘llvm::CodeModel’
                                CodeModel::Default, Ctx);
                                ^
julia_src/src/disasm.cpp:665:32: note: suggested alternatives:
In file included from julia_src/deps/srccache/llvm-svn/include/llvm/MC/MCObjectFileInfo.h:18:0,
                 from julia_src/src/disasm.cpp:41:
julia_src/deps/srccache/llvm-svn/include/llvm/Support/CodeGen.h:37:18: note:   ‘Default’
     enum Level { Default=0, Small=1, Large=2 };
                  ^
julia_src/deps/srccache/llvm-svn/include/llvm/Support/CodeGen.h:55:7: note:   ‘Default’
       Default,     // -O2, -Os
       ^
make[1]: *** [disasm.o] Error 1
make: *** [julia-src-release] Error 2

```

**Pull request merged**